File: raycast/src/lib/preview.ts
import { createReadStream } from "node:fs";
import { stat } from "node:fs/promises";
import readline from "node:readline";

import { displayTitle } from "../record";
import type { PreviewBlock, PreviewDocument, SessionRecord } from "../types";

const INJECTED_AGENTS_CONTEXT_LIMIT = 2500;

interface RecordEnvelope {
  type?: string;
  payload?: unknown;
}

interface EventPayload {
  type?: string;
  message?: string;
}

interface MessagePart {
  type?: string;
  text?: string;
}

interface OutputMessagePayload {
  type?: string;
  role?: string;
  content?: MessagePart[];
  arguments?: unknown;
  input?: unknown;
  status?: string;
  call_id?: string;
  name?: string;
  output?: string;
}

function cmdFromToolArguments(value: unknown): string {
  if (!value) {
    return "";
  }
  const tryExtract = (obj: unknown): string => {
    if (!obj || typeof obj !== "object") {
      return "";
    }
    const record = obj as Record<string, unknown>;
    const cmd = record.cmd;
    if (typeof cmd !== "string") {
      return "";
    }
    const trimmed = cmd.trim();
    if (!trimmed) {
      return "";
    }
    if (looksLikeJSONText(trimmed)) {
      try {
        const parsed = JSON.parse(trimmed) as unknown;
        const nested = tryExtract(parsed);
        if (nested) {
          return nested;
        }
      } catch {
        // ignore
      }
    }
    return trimmed;
  };

  const direct = tryExtract(value);
  if (direct) {
    return direct;
  }

  if (typeof value === "string") {
    const trimmed = value.trim();
    if (!trimmed) {
      return "";
    }
    try {
      const parsed = JSON.parse(trimmed) as unknown;
      return tryExtract(parsed);
    } catch {
      return "";
    }
  }

  return "";
}

function looksLikeJSONText(value: string): boolean {
  const trimmed = value.trim();
  return trimmed.startsWith("{") || trimmed.startsWith("[");
}

function normalizeToolOutput(value: string): string {
  const trimmed = value.trim();
  if (!trimmed || !looksLikeJSONText(trimmed)) {
    return value;
  }
  try {
    const parsed = JSON.parse(trimmed) as unknown;
    if (parsed && typeof parsed === "object" && "output" in parsed) {
      const output = (parsed as { output?: unknown }).output;
      if (typeof output === "string" && output.trim()) {
        return output.trim();
      }
    }
  } catch {
    // ignore
  }
  return value;
}

export async function parsePreviewFromFile(
  record: SessionRecord,
): Promise<PreviewDocument> {
  await stat(record.path);

  const document: PreviewDocument = {
    sessionId: record.id,
    title: displayTitle(record),
    blocks: [],
  };
  const toolNames = new Map<string, string>();

  const reader = readline.createInterface({
    input: createReadStream(record.path, { encoding: "utf8" }),
    crlfDelay: Number.POSITIVE_INFINITY,
  });

  const pushDeduped = (block: PreviewBlock) => {
    const last = document.blocks[document.blocks.length - 1];
    if (
      last &&
      last.kind === block.kind &&
      (last.title || "") === (block.title || "") &&
      (last.body || "").trim() === (block.body || "").trim()
    ) {
      return;
    }
    document.blocks.push(block);
  };

  try {
    for await (const line of reader) {
      const trimmed = line.trim();
      if (!trimmed) {
        continue;
      }

      let envelope: RecordEnvelope;
      try {
        envelope = JSON.parse(trimmed) as RecordEnvelope;
      } catch (error) {
        if (process.env.DEBUG_PREVIEW) {
          console.error(
            `[preview] Skipping malformed line in ${record.path}: ${
              error instanceof Error ? error.message : String(error)
            }`,
          );
        }
        continue;
      }

      if (envelope.type === "event_msg" || envelope.type === "event-msg") {
        const block = eventBlock(envelope.payload);
        if (block) {
          pushDeduped(block);
        }
        continue;
      }

      if (
        envelope.type === "response_item" ||
        envelope.type === "response-item"
      ) {
        for (const block of responseBlocks(envelope.payload, toolNames)) {
          pushDeduped(block);
        }
      }
    }
  } finally {
    reader.close();
  }

  if (document.blocks.length === 0) {
    document.blocks.push({
      kind: "event",
      title: "No transcript",
      body: "This session has no renderable user or assistant messages yet.",
    });
  }

  return document;
}

export function renderPreviewMarkdown(
  document: PreviewDocument,
  showSystem: boolean,
): string {
  const sections = [`# ${escapeMarkdown(document.title)}`];

  for (const block of document.blocks) {
    if (!showSystem && block.title === "Context") {
      continue;
    }

    let header = escapeMarkdown(block.title?.trim() || block.kind);
    if (block.kind === "tool_call") {
      header = `Tool: ${header}`;
    } else if (block.kind === "tool_output") {
      header = `Output: ${header}`;
    }
    const body = block.body?.trim();

    if (!body) {
      sections.push(`## ${header}`);
      continue;
    }

    if (block.kind === "tool_call") {
      let fence = looksLikeJSONText(body) ? "json" : "bash";
      if (body.includes("*** Begin Patch")) {
        fence = "diff";
      }
      sections.push(`## ${header}`, `\`\`\`${fence}`, body, "```");
      continue;
    }

    if (block.kind === "tool_output") {
      const fence = looksLikeJSONText(body) ? "json" : "text";
      sections.push(`## ${header}`, `\`\`\`${fence}`, body, "```");
      continue;
    }

    sections.push(`## ${header}`, escapeMarkdown(body));
  }

  return sections.join("\n\n");
}

export function truncate(value: string, limit: number): string {
  const trimmed = value.trim();
  if (trimmed.length <= limit) {
    return trimmed;
  }
  if (limit <= 3) {
    return trimmed.slice(0, limit);
  }
  return `${trimmed.slice(0, limit - 3)}...`;
}

export function shortenJSON(value: unknown, limit: number): string {
  if (value === undefined || value === null) {
    return "";
  }

  try {
    return truncate(JSON.stringify(value, null, 2), limit);
  } catch {
    return truncate(String(value), limit);
  }
}

function eventBlock(payload: unknown): PreviewBlock | null {
  const event = payload as EventPayload | undefined;
  switch (event?.type) {
    case "user_message":
      if (!event.message?.trim()) {
        return null;
      }
      if (isInjectedAgentsContext(event.message)) {
        return {
          kind: "event",
          title: "Context",
          body: truncate(event.message, INJECTED_AGENTS_CONTEXT_LIMIT),
        };
      }
      return { kind: "user", title: "User", body: event.message };
    case "agent_message":
      if (!event.message?.trim()) {
        return null;
      }
      if (isInjectedAgentsContext(event.message)) {
        return {
          kind: "event",
          title: "Context",
          body: truncate(event.message, INJECTED_AGENTS_CONTEXT_LIMIT),
        };
      }
      return { kind: "assistant", title: "Assistant", body: event.message };
    case "task_started":
      return null;
    case "turn/completed":
    case "turn_started":
      return { kind: "event", title: event.type };
    default:
      return null;
  }
}

function responseBlocks(
  payload: unknown,
  toolNames: Map<string, string>,
): PreviewBlock[] {
  const item = payload as OutputMessagePayload | undefined;
  switch (item?.type) {
    case "message": {
      const body = collectMessageText(item.content ?? []);
      if (!body) {
        return [];
      }
      if (item.role === "developer" || item.role === "system") {
        return [{ kind: "event", title: "Context", body }];
      }
      if (item.role === "user") {
        if (isInjectedAgentsContext(body)) {
          return [
            {
              kind: "event",
              title: "Context",
              body: truncate(body, INJECTED_AGENTS_CONTEXT_LIMIT),
            },
          ];
        }
        return [{ kind: "user", title: "User", body }];
      }
      if (isInjectedAgentsContext(body)) {
        return [
          {
            kind: "event",
            title: "Context",
            body: truncate(body, INJECTED_AGENTS_CONTEXT_LIMIT),
          },
        ];
      }
      return [{ kind: "assistant", title: "Assistant", body }];
    }
    case "function_call":
    case "tool_call":
    case "custom_tool_call": {
      const name = (item.name || "Tool Call").trim() || "Tool Call";
      const callId =
        typeof item.call_id === "string" ? item.call_id.trim() : "";
      if (callId) {
        toolNames.set(callId, name);
      }
      const cmd = cmdFromToolArguments(item.arguments);
      const input =
        typeof item.input === "string" && item.input.trim()
          ? item.input.trim()
          : "";
      const body =
        input || cmd || shortenJSON(item.input ?? item.arguments, 200);
      return [
        {
          kind: "tool_call",
          title: callId ? `${name} (${callId})` : name,
          body,
        },
      ];
    }
    case "function_call_output":
    case "tool_call_output":
    case "tool_output":
    case "custom_tool_call_output": {
      const callId =
        typeof item.call_id === "string" ? item.call_id.trim() : "";
      const name = callId ? toolNames.get(callId) : undefined;
      const output = normalizeToolOutput(
        item.output?.trim() || "command completed",
      );
      return [
        {
          kind: "tool_output",
          title:
            callId && name ? `${name} (${callId})` : callId || "Tool Output",
          body: truncate(output, 500),
        },
      ];
    }
    default:
      return [];
  }
}

function collectMessageText(parts: MessagePart[]): string {
  return parts
    .filter(
      (part) =>
        part.text &&
        (part.type === "input_text" ||
          part.type === "output_text" ||
          part.type === "text"),
    )
    .map((part) => part.text?.trim() || "")
    .filter(Boolean)
    .join("\n")
    .trim();
}

function escapeMarkdown(value: string): string {
  return value.replace(/\r\n/g, "\n").replace(/\r/g, "\n").trim();
}

function isInjectedAgentsContext(message: string): boolean {
  const trimmed = message.trim();
  if (!trimmed) {
    return false;
  }

  const firstLine = (trimmed.split("\n", 1)[0] || "").trim();
  const lower = trimmed.toLowerCase();
  const lowerFirst = firstLine.toLowerCase();

  const mentionsAgents =
    lowerFirst === "agents.md" ||
    lowerFirst.startsWith("# agents.md") ||
    lowerFirst.includes("agents.md instructions for");
  if (!mentionsAgents) {
    return false;
  }

  return (
    lower.includes("<instructions>") || lower.includes("<environment_context>")
  );
}

File: internal/session/preview.go
package session

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"sync"

	"github.com/charmbracelet/x/ansi"
	"github.com/muesli/reflow/wordwrap"
)

type cachedPreview struct {
	modTime int64
	doc     PreviewDocument
}

// PreviewCache caches parsed previews by path and mtime.
type PreviewCache struct {
	mu    sync.Mutex
	items map[string]cachedPreview
}

// NewPreviewCache creates an in-memory preview cache.
func NewPreviewCache() *PreviewCache {
	return &PreviewCache{items: make(map[string]cachedPreview)}
}

// Load returns the cached preview if present or parses the rollout.
func (c *PreviewCache) Load(record SessionRecord) (PreviewDocument, error) {
	stat, err := os.Stat(record.Path)
	if err != nil {
		return PreviewDocument{}, fmt.Errorf("stat preview %s: %w", record.Path, err)
	}
	key := record.Path
	modTime := stat.ModTime().UnixNano()

	c.mu.Lock()
	if cached, ok := c.items[key]; ok && cached.modTime == modTime {
		c.mu.Unlock()
		return cached.doc, nil
	}
	c.mu.Unlock()

	doc, err := parsePreview(record)
	if err != nil {
		return PreviewDocument{}, fmt.Errorf("parse preview %s: %w", record.Path, err)
	}

	c.mu.Lock()
	c.items[key] = cachedPreview{modTime: modTime, doc: doc}
	c.mu.Unlock()
	return doc, nil
}

func parsePreview(record SessionRecord) (PreviewDocument, error) {
	file, err := os.Open(record.Path)
	if err != nil {
		return PreviewDocument{}, fmt.Errorf("open preview %s: %w", record.Path, err)
	}
	defer file.Close()

	doc := PreviewDocument{
		SessionID: record.ID,
		Title:     record.DisplayTitle(),
	}
	toolNames := make(map[string]string)
	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
	for scanner.Scan() {
		var env recordEnvelope
		if err := json.Unmarshal(scanner.Bytes(), &env); err != nil {
			continue
		}
		switch env.Type {
		case RecordTypeEventMsg, RecordTypeEventMsgHyphen:
			if block := eventBlock(env.Payload); block != nil {
				doc.Blocks = appendBlockDedup(doc.Blocks, *block)
			}
		case RecordTypeResponseItem, RecordTypeResponseItemHyphen:
			if blocks := responseBlocks(env.Payload, toolNames); len(blocks) > 0 {
				for _, block := range blocks {
					doc.Blocks = appendBlockDedup(doc.Blocks, block)
				}
			}
		}
	}
	if err := scanner.Err(); err != nil {
		return PreviewDocument{}, fmt.Errorf("scan preview %s: %w", record.Path, err)
	}
	if len(doc.Blocks) == 0 {
		doc.Blocks = append(doc.Blocks, PreviewBlock{
			Kind:  PreviewEvent,
			Title: "No transcript",
			Body:  "This session has no renderable user or assistant messages yet.",
		})
	}
	return doc, nil
}

func appendBlockDedup(blocks []PreviewBlock, next PreviewBlock) []PreviewBlock {
	if len(blocks) == 0 {
		return append(blocks, next)
	}
	last := blocks[len(blocks)-1]
	if last.Kind == next.Kind && last.Title == next.Title && strings.TrimSpace(last.Body) == strings.TrimSpace(next.Body) {
		return blocks
	}
	return append(blocks, next)
}

func eventBlock(raw json.RawMessage) *PreviewBlock {
	var payload eventPayload
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil
	}
	switch payload.Type {
	case "user_message":
		if strings.TrimSpace(payload.Message) == "" {
			return nil
		}
		if isInjectedAgentsContext(payload.Message) {
			return &PreviewBlock{Kind: PreviewEvent, Title: "Context", Body: truncate(payload.Message, injectedAgentsContextPreviewLimit)}
		}
		return &PreviewBlock{Kind: PreviewUser, Title: "User", Body: payload.Message}
	case "agent_message":
		if strings.TrimSpace(payload.Message) == "" {
			return nil
		}
		if isInjectedAgentsContext(payload.Message) {
			return &PreviewBlock{Kind: PreviewEvent, Title: "Context", Body: truncate(payload.Message, injectedAgentsContextPreviewLimit)}
		}
		return &PreviewBlock{Kind: PreviewAssistant, Title: "Assistant", Body: payload.Message}
	case "task_started":
		return nil
	case "turn/completed", "turn_started":
		return &PreviewBlock{Kind: PreviewEvent, Title: payload.Type, Body: ""}
	default:
		return nil
	}
}

func cmdFromToolArgs(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}

	type cmdCarrier struct {
		Cmd string `json:"cmd"`
	}

	var obj cmdCarrier
	if err := json.Unmarshal(raw, &obj); err == nil && strings.TrimSpace(obj.Cmd) != "" {
		cmd := strings.TrimSpace(obj.Cmd)
		if looksLikeJSONText(cmd) {
			var nested cmdCarrier
			if json.Unmarshal([]byte(cmd), &nested) == nil && strings.TrimSpace(nested.Cmd) != "" {
				return strings.TrimSpace(nested.Cmd)
			}
		}
		return cmd
	}

	var asString string
	if err := json.Unmarshal(raw, &asString); err == nil {
		asString = strings.TrimSpace(asString)
		if asString == "" {
			return ""
		}
		if json.Unmarshal([]byte(asString), &obj) == nil && strings.TrimSpace(obj.Cmd) != "" {
			cmd := strings.TrimSpace(obj.Cmd)
			if looksLikeJSONText(cmd) {
				var nested cmdCarrier
				if json.Unmarshal([]byte(cmd), &nested) == nil && strings.TrimSpace(nested.Cmd) != "" {
					return strings.TrimSpace(nested.Cmd)
				}
			}
			return cmd
		}
	}
	return ""
}

func stringFromRaw(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	var asString string
	if err := json.Unmarshal(raw, &asString); err == nil {
		return strings.TrimSpace(asString)
	}
	return strings.TrimSpace(string(raw))
}

func normalizeToolOutput(output string) string {
	trimmed := strings.TrimSpace(output)
	if trimmed == "" {
		return output
	}
	if !looksLikeJSONText(trimmed) {
		return output
	}
	var carrier struct {
		Output string `json:"output"`
	}
	if json.Unmarshal([]byte(trimmed), &carrier) == nil && strings.TrimSpace(carrier.Output) != "" {
		return strings.TrimSpace(carrier.Output)
	}
	return output
}

func responseBlocks(raw json.RawMessage, toolNames map[string]string) []PreviewBlock {
	var payload outputMessagePayload
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil
	}
	switch payload.Type {
	case "message":
		text := collectMessageText(payload.Content)
		if text == "" {
			text = strings.TrimSpace(payload.Message)
		}
		if text == "" {
			return nil
		}
		switch payload.Role {
		case "developer", "system":
			return []PreviewBlock{{Kind: PreviewEvent, Title: "Context", Body: text}}
		case "user":
			if isInjectedAgentsContext(text) {
				return []PreviewBlock{{Kind: PreviewEvent, Title: "Context", Body: truncate(text, injectedAgentsContextPreviewLimit)}}
			}
			return []PreviewBlock{{Kind: PreviewUser, Title: "User", Body: text}}
		default:
			if isInjectedAgentsContext(text) {
				return []PreviewBlock{{Kind: PreviewEvent, Title: "Context", Body: truncate(text, injectedAgentsContextPreviewLimit)}}
			}
			return []PreviewBlock{{Kind: PreviewAssistant, Title: "Assistant", Body: text}}
		}
	case "function_call", "tool_call", "custom_tool_call":
		body := stringFromRaw(payload.Input)
		if body == "" {
			body = cmdFromToolArgs(payload.Args)
		}
		if body == "" {
			if strings.TrimSpace(payload.Name) == "exec_command" {
				body = shortenJSON(payload.Args, 200)
			} else if len(payload.Input) > 0 {
				body = shortenJSON(payload.Input, 400)
			} else {
				body = shortenJSON(payload.Args, 200)
			}
		}
		name := strings.TrimSpace(payload.Name)
		if name == "" {
			name = "Tool Call"
		}
		title := name
		if payload.CallID != "" {
			toolNames[payload.CallID] = name
			title = fmt.Sprintf("%s (%s)", name, payload.CallID)
		}
		return []PreviewBlock{{Kind: PreviewToolCall, Title: title, Body: body}}
	case "function_call_output", "tool_call_output", "tool_output", "custom_tool_call_output":
		body := normalizeToolOutput(payload.Output)
		if body == "" {
			body = "command completed"
		}
		callID := strings.TrimSpace(payload.CallID)
		title := callID
		if callID == "" {
			title = "Tool Output"
		} else if name := strings.TrimSpace(toolNames[callID]); name != "" {
			title = fmt.Sprintf("%s (%s)", name, callID)
		}
		return []PreviewBlock{{Kind: PreviewToolOut, Title: title, Body: truncate(body, 500)}}
	default:
		return nil
	}
}

func collectMessageText(parts []messagePart) string {
	chunks := make([]string, 0, len(parts))
	for _, part := range parts {
		if part.Text == "" {
			continue
		}
		if part.Type == "input_text" || part.Type == "output_text" || part.Type == "text" {
			chunks = append(chunks, part.Text)
		}
	}
	return strings.TrimSpace(strings.Join(chunks, "\n"))
}

func shortenJSON(raw json.RawMessage, limit int) string {
	if len(raw) == 0 {
		return ""
	}
	var out bytes.Buffer
	if err := json.Indent(&out, raw, "", "  "); err == nil {
		return truncate(out.String(), limit)
	}
	return truncate(string(raw), limit)
}

func truncate(s string, limit int) string {
	s = strings.TrimSpace(s)
	if len(s) <= limit {
		return s
	}
	if limit <= 3 {
		return s[:limit]
	}
	return s[:limit-3] + "..."
}

func codeFence(lang, content string) string {
	content = strings.TrimSpace(content)
	if content == "" {
		if lang == "" {
			return "```\n```"
		}
		return "```" + lang + "\n```"
	}
	maxRun := 0
	run := 0
	for _, r := range content {
		if r == '`' {
			run++
			continue
		}
		if run > maxRun {
			maxRun = run
		}
		run = 0
	}
	if run > maxRun {
		maxRun = run
	}
	fenceLen := 3
	if maxRun >= fenceLen {
		fenceLen = maxRun + 1
	}
	fence := strings.Repeat("`", fenceLen)
	if lang != "" {
		return fmt.Sprintf("%s%s\n%s\n%s", fence, lang, content, fence)
	}
	return fmt.Sprintf("%s\n%s\n%s", fence, content, fence)
}

func padRightANSI(s string, width int) string {
	s = ansi.Truncate(s, width, "")
	pad := width - ansi.StringWidth(s)
	if pad <= 0 {
		return s
	}
	return s + strings.Repeat(" ", pad)
}

func looksLikeJSONText(s string) bool {
	trimmed := strings.TrimSpace(s)
	if trimmed == "" {
		return false
	}
	return strings.HasPrefix(trimmed, "{") || strings.HasPrefix(trimmed, "[")
}

func renderBubble(label, content string, outerWidth, leftPad int) string {
	if outerWidth < 20 {
		outerWidth = 20
	}
	if leftPad < 0 {
		leftPad = 0
	}
	innerWidth := outerWidth - 2
	if innerWidth < 8 {
		innerWidth = 8
		outerWidth = innerWidth + 2
	}

	label = strings.TrimSpace(label)
	if label == "" {
		label = "Message"
	}

	labelMax := outerWidth - 5
	if labelMax < 0 {
		labelMax = 0
	}
	if ansi.StringWidth(label) > labelMax {
		label = ansi.Truncate(label, labelMax, "…")
	}
	dashLen := outerWidth - (ansi.StringWidth(label) + 5)
	if dashLen < 0 {
		dashLen = 0
	}

	prefix := strings.Repeat(" ", leftPad)
	lines := make([]string, 0, 4)
	lines = append(lines, prefix+"╭─ "+label+" "+strings.Repeat("─", dashLen)+"╮")

	content = strings.TrimRight(content, "\n")
	bodyLines := []string{""}
	if strings.TrimSpace(content) != "" {
		bodyLines = strings.Split(content, "\n")
	}
	for _, line := range bodyLines {
		line = padRightANSI(line, innerWidth)
		lines = append(lines, prefix+"│"+line+"│")
	}
	lines = append(lines, prefix+"╰"+strings.Repeat("─", outerWidth-2)+"╯")
	return strings.Join(lines, "\n")
}

func computeBubbleLayout(width int) (availableWidth, bubbleWidth int) {
	availableWidth = width
	if availableWidth > 2 {
		availableWidth -= 2
	}
	if availableWidth < 20 {
		availableWidth = 20
	}

	bubbleWidth = int(float64(availableWidth) * 0.92)
	if bubbleWidth < 20 {
		bubbleWidth = 20
	}
	if bubbleWidth > availableWidth {
		bubbleWidth = availableWidth
	}
	return availableWidth, bubbleWidth
}

func filterPreviewBlocks(blocks []PreviewBlock, showSystem bool) []PreviewBlock {
	if showSystem {
		return blocks
	}
	out := make([]PreviewBlock, 0, len(blocks))
	for _, block := range blocks {
		if block.Title == "Context" {
			continue
		}
		out = append(out, block)
	}
	return out
}

func bubbleLabelAndLang(kind PreviewBlockKind, header, body string) (string, string) {
	switch kind {
	case PreviewToolCall:
		lowerHeader := strings.ToLower(header)
		if strings.Contains(body, "*** Begin Patch") || strings.HasPrefix(lowerHeader, "apply_patch") {
			return "Tool: " + header, "diff"
		}
		if looksLikeJSONText(body) {
			return "Tool: " + header, "json"
		}
		return "Tool: " + header, "bash"
	case PreviewToolOut:
		if looksLikeJSONText(body) {
			return "Output: " + header, "json"
		}
		return "Output: " + header, "text"
	case PreviewEvent:
		if header == "Context" {
			return header, "text"
		}
		return header, ""
	default:
		return header, ""
	}
}

func renderBlockBody(body, lang string, mr *MarkdownRenderer, width int) string {
	if strings.TrimSpace(body) == "" {
		return ""
	}
	if mr == nil {
		return wordwrap.String(body, width)
	}
	markdown := body
	if lang != "" {
		markdown = codeFence(lang, body)
	}
	rendered, err := mr.Render(markdown, width)
	if err != nil {
		return wordwrap.String(body, width)
	}
	return rendered
}

func flattenOneLine(s string) (string, bool) {
	trimmed := strings.TrimSpace(s)
	if trimmed == "" {
		return "", false
	}
	hasNewline := strings.Contains(trimmed, "\n")
	parts := strings.Fields(trimmed)
	if len(parts) == 0 {
		return "", hasNewline
	}
	return strings.Join(parts, " "), hasNewline
}

func stripCallIDSuffix(title string) string {
	title = strings.TrimSpace(title)
	if title == "" {
		return ""
	}
	if idx := strings.LastIndex(title, " (call"); idx > 0 && strings.HasSuffix(title, ")") {
		return strings.TrimSpace(title[:idx])
	}
	return title
}

func execCommandOutputBody(output string) string {
	const marker = "\nOutput:\n"
	if idx := strings.Index(output, marker); idx >= 0 {
		if body := strings.TrimSpace(output[idx+len(marker):]); body != "" {
			return body
		}
	}
	return output
}

func renderToolLine(kind PreviewBlockKind, title, body string, width int) string {
	if width < 20 {
		width = 20
	}
	if kind == PreviewToolCall && looksLikeJSONText(body) && strings.Contains(body, "\"cmd\"") {
		if cmd := cmdFromToolArgs(json.RawMessage(body)); cmd != "" {
			body = cmd
		}
	}
	title = stripCallIDSuffix(title)
	if strings.EqualFold(strings.TrimSpace(title), "exec_command") {
		title = ""
	}
	if kind == PreviewToolOut && title == "" {
		body = execCommandOutputBody(body)
	}
	flat, multiline := flattenOneLine(body)
	if flat == "" {
		flat = "…"
	}

	prefix := "> "
	if kind == PreviewToolOut {
		prefix = "< "
	}

	sep := ": "
	if strings.TrimSpace(title) == "" {
		sep = ""
	}
	minCmd := 10
	maxTitleWidth := width - ansi.StringWidth(prefix) - ansi.StringWidth(sep) - minCmd
	if maxTitleWidth <= 0 {
		title = ""
		sep = ""
	} else if ansi.StringWidth(title) > maxTitleWidth {
		title = ansi.Truncate(title, maxTitleWidth, "…")
	}

	remainingCmd := width - ansi.StringWidth(prefix) - ansi.StringWidth(title) - ansi.StringWidth(sep)
	if remainingCmd < 0 {
		remainingCmd = 0
	}
	cmdPart := ansi.Truncate(flat, remainingCmd, "…")
	line := prefix + title + sep + cmdPart
	if multiline && remainingCmd >= 2 {
		line = ansi.Truncate(line+" …", width, "…")
	}
	return ansi.Truncate(line, width, "…")
}

// RenderPreview returns a wrapped preview suitable for a viewport.
// If mr is provided, individual messages will be rendered as markdown with ANSI formatting.
func RenderPreview(doc PreviewDocument, width int, showSystem bool, mr *MarkdownRenderer) string {
	if width < 20 {
		width = 20
	}

	parts := make([]string, 0, len(doc.Blocks)+2)
	parts = append(parts, strings.TrimSpace(doc.Title))

	blocks := filterPreviewBlocks(doc.Blocks, showSystem)

	availableWidth, bubbleWidth := computeBubbleLayout(width)

	for _, block := range blocks {
		header := strings.TrimSpace(block.Title)
		if header == "" {
			header = string(block.Kind)
		}
		body := strings.TrimSpace(block.Body)

		if block.Kind == PreviewToolCall || block.Kind == PreviewToolOut {
			parts = append(parts, renderToolLine(block.Kind, header, body, width))
			continue
		}

		label, lang := bubbleLabelAndLang(block.Kind, header, body)

		outerWidth := bubbleWidth
		leftPad := 0
		if block.Kind == PreviewUser {
			leftPad = availableWidth - bubbleWidth
			if leftPad < 0 {
				leftPad = 0
			}
		}
		innerWidth := outerWidth - 2

		content := renderBlockBody(body, lang, mr, innerWidth)

		parts = append(parts, renderBubble(label, content, outerWidth, leftPad))
	}
	return strings.Join(parts, "\n\n")
}
